equinox/nn/_attention.py

```diff
@@ -325,6 +401,42 @@ def __call__(
                 raise ValueError(
                     "process_heads must not change the shape of the heads."
                 )
+        state_length = None
```

Comment:
By the way, prefer single-assignment-form here:
```python
if state is None:
    state_length = None
    causal_mask_offset = 0
else:
    ...
    state_length = ...
    ...
```
which is generally a more reliable coding style than overwriting a local variable.